### ISSUE TITLE
how should initialization of a Replicated array behave?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#15731]
Consider an array that is declared to be dmapped with ReplicatedDist (see https://chapel-lang.org/docs/modules/dists/ReplicatedDist.html ).

For example:

``` chapel
    var D = {1..1} dmapped Replicated();
    var A: [D] MyRecord = makeRecord();
```

How should `A` be initialized? Supposing that there are 2 locales, the array A declared above has the following elements:
 * A[1] on Locale 0
 * A[1] on Locale 1

The ReplicatedDist docs say this:

> Consistency among the replicands is not preserved automatically. That is, changes to one replicand of an array are never propagated to the other replicands by the distribution implementation. If desired, consistency needs to be maintained by the user.

However, currently (after PR #15239), the value on Locale 1 is simply never initialized. It is later deinitialized (which is a memory error).

How could we fix this?
 1. ReplicatedDist arrays are always default initialized and then assigned during initialization, so that all elements not initialized will have a default value. 
 2. ReplicatedDist arrays copy elements that are initialized to all the replicands
 3. ReplicatedDist arrays default-initialize the other replicands corresponding to the elements that were initialized. (The problem with this approach is that it might depend on the initialization expression which replicands were initialized)
 4. ReplicatedDist arrays track which elements are initialized, raise a runtime error if an uninitialized element is accessed, and only deinitialized initialized elements

**Comment by mppf** [comment]
PR #15732 implements option 1 for the near term.

**Comment by mppf** [comment]
@vasslitvinov @benharsh - do you have any thoughts here?

**Comment by vasslitvinov** [comment]
This is a tough one. I support (1) for the time being.

Thinking a bit forward, what if the user actually wants to create a Replicated array, initializing all locales explicitly?

Replicated has some scatter-like functionality, or we can introduce one, that the user can apply to indicate explicitly "I want to initialize all locales". We probably would have to make a special case in the compiler for that.

Alternatively, Replicated could provide a factory function that takes a single-locale array and produces a Replicated array with copies of that.

**Comment by bradcray** [comment]
Option 3 seems most intuitive to me.  What are its downsides?

**Comment by mppf** [comment]
@bradcray 

> > 3. ReplicatedDist arrays copy elements that are initialized to all the replicands

> Option 3 seems most intuitive to me. What are its downsides?

The main issue is that I'm not sure how ReplicatedDist would be able to identify the initialization pattern. In particular, the RHS initialization expression could be a `for` or `forall` over an arbitrary type and so could be evaluated on arbitrary locales. Perhaps this just means that we'd have to add to the dsi interface something like `dsiInitializedElement` that for replicated would run the initializers for all the replicands - but in my work so far I was trying to keep the dsi changes pretty minimal. Part of the reason for that is that I wanted bulk transfer to work when initializing array elements (but see also #15675 / #15676). Anyway, it seems like with careful dsi interface design we could allow that & also allow Replicated to have the right hook to do this.

So another way to put all of that is that the main downside I see is "implementation complexity".

**Comment by vasslitvinov** [comment]
To me the downside of Option 3 is inconsistent semantics. My understanding is that a Replicated array is consistently presented to the user as living on the current locale, as far as the standard array operations. Option 3 makes initialization behave differently.

Implementation-wise, I am not following Michael's implementation approach above. To me it seems like it would be a two-step process: (a) initialize the current replicand as we would a DR array on the current locale, and (b) copy that to each other locale. For each element, (b) would use the copy initializer (semantically speaking), like initializing a variable on a remote locale from a variable on the current locale.

**Comment by mppf** [comment]
> Implementation-wise, I am not following Michael's implementation approach above. To me it seems like it would be a two-step process: (a) initialize the current replicand as we would a DR array on the current locale, and (b) copy that to each other locale. For each element, (b) would use the copy initializer (semantically speaking), like initializing a variable on a remote locale from a variable on the current locale.

I would agree with what you said above - my comment had to do with how to implement that (mainly, figuring out which replicand was initialized and we should copy from).

**Comment by bradcray** [comment]
Based on Vass's hesitancy with the semantics of option 3 (which I agree-ish with), my next intuition would be:

* the initialization expression is evaluated and assigned to the local replicand
* all other replicands are default initialized
* if default initialization isn't supported by the type, it's not legal (seems unfortunate, but...?)

I think this is closest to item 4, though I'm not understanding the comment relating to trying to figure out which replicand is initialized.  Specifically, for an n-element replicated array, I think it can only be initialized locally where the task is running and that the initialization expression can only generate n elements (like any other array initialization).  Maybe put another way, I'm not aware of a way that one could write a closed-form initialization expression for a replicated array that would initialize numLocales * n elements, nor do I think we should try to support that through normal initialization expressions.  In a comment above, Michael says:

> In particular, the RHS initialization expression could be a for or forall over an arbitrary type and so could be evaluated on arbitrary locales. 

but I don't think that this pattern would affect which elements or replicands are initialized, only where those initial values are computed.  For example, for a normal array, I could say:

```chapel
var A: [1..3] int = for loc in Locales[0..2] do on loc do here.id;
```

but I'm only the initializing local elements of A on the local locale despite the fact that I'm using an on-clause to compute the values (needlessly in this example).  So similarly, I'd expect that initialization of a Replicated array would be the same, assuming we didn't go with option 3: we evaluate the expression wherever it wants to be, but we only initialize the n local elements.

**Comment by mppf** [comment]
> but I'm only the initializing local elements of A on the local locale despite the fact that I'm using an on-clause to compute the values (needlessly in this example). So similarly, I'd expect that initialization of a Replicated array would be the same, assuming we didn't go with option 3: we evaluate the expression wherever it wants to be, but we only initialize the n local elements.

Ah, you've corrected a mistake I was making in my thinking, thanks for that. (at the very low level view, chpl__transferArray has e.g. `[ (aa,bb) in zip(a,b) ] initialize-aa-from-bb` where `a` is the LHS array; so the locale running the initialization is determined by the array and not by the initialization expression).

**Comment by vasslitvinov** [comment]
[Brad's suggested rules above](https://github.com/chapel-lang/chapel/issues/15731#issuecomment-635569307) sound good to me.

Option 1 is a simpler-to-implement, more restrictive version of that (clearly). I am OK keeping it for now.

**Comment by vasslitvinov** [comment]
I suggest that if RHS is a Replicated array then ALL replicands get copied over upon initialization or assignment. [Perhaps this should go in its own issue?] In particular, initialization like the following:
```chpl
var ra2 = ra1;
```
today produces different values of `ra2` depending on whether copy elision kicks in:
* `ra1` is "dead" => `ra1` is MOVED into `ra2` => all replicands are initialized
* otherwise => `ra1` is COPIED into `ra2` ==> only the current replicand of `ra2` is initialized from `ra1`, the rest are default-initialized

Assignment `ra2 = ra1;` today always copies just the current replicand.

The above behavior is mostly consistent with the view "a replicated array behaves as if it is just the current replicand." However it may surprise users who expect `var ra2 = ra1;` to do a copy on all locales. Especially if the result is that `ra2` is a replicated array.

Alternatively, `var ra2 = ra1;` could result in a non-distributed array that is a copy of ra1's current replicand Analogously,
```chpl
proc f(in A) { }
var ra = (a Replicated array);
f(ra);  // the formal A is a default rectangular array
```
